qadence/analog/addressing.py

```diff
@@ -0,0 +1,138 @@
+    max_det: float | torch.Tensor | Parameter = DEFAULT_MAX_DETUNING
+
+    # weight constraint
+    weight_constraint: WeightConstraint = WeightConstraint.NORMALIZE
+
+    def _normalize_weights(self) -> None:
+        self.weights_amp = {
+            k: Parameter(v) if not isinstance(v, Parameter) else abs(v)
+            for k, v in self.weights_amp.items()
+        }
+        sum_weights_amp = sum(list(self.weights_amp.values()))
+        self.weights_amp = {k: v / sum_weights_amp for k, v in self.weights_amp.items()}
+
+        self.weights_det = {
+            k: Parameter(v) if not isinstance(v, Parameter) else abs(v)
+            for k, v in self.weights_det.items()
+        }
+        sum_weights_det = sum(list(self.weights_det.values()))
+        self.weights_det = {k: v / sum_weights_det for k, v in self.weights_det.items()}
+
+    def _restrict_weights(self) -> None:
+        self.weights_amp = {
+            k: v * (sigmoid(v, 20, 0.0) - sigmoid(v, 20.0, -1.0))
+            for k, v in self.weights_amp.items()
+        }
+        self.weights_det = {
+            k: v * (sigmoid(v, 20.0, 0.0) - sigmoid(v, 20.0, -1.0))
+            for k, v in self.weights_det.items()
+        }
+
+    def _restrict_max_vals(self) -> None:
+        self.max_amp = self.max_amp * (
+            sympy.Heaviside(self.max_amp) - sympy.Heaviside(self.max_amp - DEFAULT_MAX_AMPLITUDE)
+        )
+        self.max_det = self.max_det * (
+            sympy.Heaviside(self.max_det) - sympy.Heaviside(self.max_det - DEFAULT_MAX_DETUNING)
+        )
+
+    def __post_init__(self) -> None:
+        # validate weights
+        if all([not isinstance(v, Parameter) for v in self.weights_amp.values()]):
+            if not torch.isclose(
+                torch.tensor(list(self.weights_amp.values())).sum(),
+                torch.tensor(1.0),
+                atol=1e-3,
+            ):
+                raise ValueError("Amplitude addressing pattern weights must sum to 1.0")
+        if all([not isinstance(v, Parameter) for v in self.weights_det.values()]):
+            if not torch.isclose(
+                torch.tensor(list(self.weights_det.values())).sum(),
+                torch.tensor(1.0),
+                atol=1e-3,
+            ):
+                raise ValueError("Detuning addressing pattern weights must sum to 1.0")
+
+        # validate detuning value
+        if not isinstance(self.max_amp, Parameter):
+            if self.max_amp > DEFAULT_MAX_AMPLITUDE:
+                warn("Maximum absolute value of amplitude is exceeded")
+        if not isinstance(self.max_det, Parameter):
+            if self.max_det > DEFAULT_MAX_DETUNING:
+                warn("Maximum absolute value of detuning is exceeded")
+
+        # augment weight dicts if needed
+        self.weights_amp = {
+            i: Parameter(0.0) if i not in self.weights_amp else self.weights_amp[i]
+            for i in range(self.n_qubits)
+        }
```

Comment:
What if the other parameters are not variational? You would have some fixed weights and some variational weights mixed up, I guess, right? I would just use the same approach for all the weights.

Reply:
Yes, some mixing can happen that might be confusing. I can write some checks that enforce either all weights to be constant or all - variational. If user passes a one or more constant values as weights that are between 0 and 1 - everything is ok and all other weights are set constant and equal to 0. If one or more weights are passed as variational parameters - the rest of parameters are also set constant, thus not trainable. Thus all weights will be trainable only if the user explicitly passes weights for each qubit.

Reply:
I don't know much here but perhaps you can convert it directly in this passage ?
```suggestion
        has_parameter = np.any([isinstance(Parameter, self.weights_amp[i] for i in range(self.n_qubits)])
        for i in range(self.n_qubits):
            if i not in self.weights_amp:
                self.weights_amp[i] = (Parameter(0.0) if has_parameter else 0.0) 
            elif has_parameter and not isinstance(Parameter, self.weights_amp[i]):
                self.weights_amp[i] = Parameter(self.weights_amp[i])
```